scripts/relup-base-vsns.escript

```diff
@@ -0,0 +1,269 @@
+#!/usr/bin/env escript
+%% -*- mode: erlang; -*-
+
+-mode(compile).
+
+-define(RED, "\e[31m").
+-define(RESET, "\e[39m").
+
+usage() ->
+"A script to manage the released versions of EMQX for relup and hot
+upgrade/downgrade.
+
+We store a \"database\" of released versions as an `eterm' file, which
+is a mapping from a given version `Vsn' to its OTP version and a list
+of previous versions from which one can upgrade to `Vsn' (the
+\"from_versions\" list).  That allow us to more easily/explicitly keep
+track of allowed version upgrades/downgrades, as well as OTP changes
+between releases
+
+In the examples below, `VERSION_DB_PATH' represents the path to the
+`eterm' file containing the version database to be used.
+
+Usage:
+
+  * List the previous base versions from which `TO_VSN' may be
+    upgraded to.  Used to list versions for which relup files are to
+    be made.
+
+    relup-base-vsns.escript base-vsns TO_VSN VERSION_DB_PATH
+
+
+  * Show the OTP version with which `Vsn' was built.
+
+    relup-base-vsns.escript otp-vsn-for VSN VERSION_DB_PATH
+
+
+  * Automatically inserts a new version into the database.  Previous
+    versions with the same Major and Minor numbers as `Vsn' are
+    considered to be upgradeable from, and versions with higher Major
+    and Minor numbers will automatically include `Vsn' in their
+    \"from_versions\" list.
+
+    For example, if inserting 4.4.8 when 4.5.0 and 4.5.1 exists,
+    versions `BASE_FROM_VSN'...4.4.7 will be considered 4.4.8's
+    \"from_versions\", and 4.4.8 will be included into 4.5.0 and
+    4.5.1's from versions.
+
+    relup-base-vsns.escript insert-new-vsn NEW_VSN BASE_FROM_VSN OTP_VSN VERSION_DB_PATH
+
+  * Check if the version database is consistent considering `VSN'.
+
+    relup-base-vsns.escript check-vsn-db VSN VERSION_DB_PATH
+".
+
+main(["base-vsns", To0, VsnDB]) ->
+    {ok, [VsnMap]} = file:consult(VsnDB),
+    To = strip_pre_release(To0),
```

Comment:
these two lines seem to be repeating for each clause
consider moving to a function which does both
e.g. `read_db(VsnDB)`

Reply:
They do different unrelated things, maybe it's strange to mix them up?
One just normalizes user input, the other only loads the eterm file.